test/integration/instance.validations.test.ts

```diff
@@ -0,0 +1,978 @@
+        },
+      },
+    }, { sequelize });
+
+    await User.sync();
+    const error = await expect(User.build({ name: 'error' }).validate()).to.be.rejected;
+    expect(error).to.be.instanceof(ValidationError);
+    if (error instanceof ValidationError) {
+      expect(error.get('name')[0].message).to.equal('Invalid username');
+    }
+
+    await expect(User.build({ name: 'no error' }).validate()).not.to.be.rejected;
+  });
+
+  it('skips other validations if allowNull is true and the value is null', async () => {
+    class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      declare age: number;
+    }
+
+    User.init({
+      age: {
+        type: DataTypes.INTEGER,
+        allowNull: true,
+        validate: {
+          min: {
+            args: [0],
+            msg: 'must be positive',
+          },
+        },
+      },
+    }, { sequelize });
+
+    const error = await expect(User
+      .build({ age: -1 })
+      .validate())
+      .to.be.rejected;
+
+    expect(error.get('age')[0].message).to.equal('must be positive');
+  });
+
+  it('validates a model with custom model-wide validation methods', async () => {
+    class Foo extends Model<InferAttributes<Foo>, InferCreationAttributes<Foo>> {
+      declare field1: number | null;
+      declare field2: number | null;
+    }
+
+    Foo.init({
+      field1: {
+        type: DataTypes.INTEGER,
+        allowNull: true,
+      },
+      field2: {
+        type: DataTypes.INTEGER,
+        allowNull: true,
+      },
+    }, {
+      sequelize,
+      validate: {
+        xnor() {
+          if (this.field1 === null === (this.field2 === null)) {
+            throw new Error('xnor failed');
+          }
+        },
+      },
+    });
+
+    const error = await expect(Foo
+      .build({ field1: null, field2: null })
+      .validate())
+      .to.be.rejected;
+
+    expect(error.get('xnor')[0].message).to.equal('xnor failed');
+
+    await expect(Foo
+      .build({ field1: 33, field2: null })
+      .validate())
+      .not.to.be.rejected;
+  });
+
+  it('validates model with a validator whose arg is an Array successfully twice in a row', async () => {
+    class Foo extends Model<InferAttributes<Foo>, InferCreationAttributes<Foo>> {
+      declare bar: string;
+    }
+
+    Foo.init({
+      bar: {
+        type: DataTypes.STRING,
+        validate: {
+          isIn: [['a', 'b']],
+        },
+      },
+    }, {
+      sequelize,
+    });
+
+    const foo = Foo.build({ bar: 'a' });
+    await expect(foo.validate()).not.to.be.rejected;
+    await expect(foo.validate()).not.to.be.rejected;
+  });
+
+  it('validates enums', async () => {
+    enum TestEnum {
+      value1 = 'value1',
+      value2 = 'value2',
+    }
+    enum FailingTestEnum {
+      value3 = 'value3',
+    }
+    const values = Object.values(TestEnum);
+
+    class Bar extends Model<InferAttributes<Bar>, InferCreationAttributes<Bar>> {
+      declare field: TestEnum | FailingTestEnum;
```

Comment:
`FailingTestEnum` is not part of this model's definition

```suggestion
      declare field: TestEnum;
```

When you try to do this:

```
const failingBar = Bar.build({ field: FailingTestEnum.value3 });
```

Simply use `// @ts-expect-error -- testing that invalid enum values get rejected` to tell typescript that that is expected to fail

Reply:
@ephys Thanks for teaching new lesson about TypeScript, I will use my TypeScript knowledge to improve my contribution to Sequelize.